### PR TITLE
Path paint

### PR DESCRIPTION
**Pull request by marijnvdwerf**
image

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
@duncanspumpkin Base function is mostly done at this point (except for the nearly identical branch)

**Comment by RobertWHurst** [comment]
Aren't the concrete paths with the wooden supports missing their railings?

**Comment by marijnvdwerf** [comment]
@RobertWHurst Hence the 'WIP'

**Comment by RobertWHurst** [comment]
@marijnvdwerf Ah, gotcha :)

File: src/interface/paint_path.c [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+                    break;
+            }
+        }
+
+        if (!(map_element->properties.path.type & 0x08)) {
+            return;
+        }
+
+        // Draw ride sign
+        RCT2_CALLPROC_X(0x006A4A71, 0, 0, get_current_rotation(), height, (int) map_element, 0, ebp);
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: @IntelOrca You did the other ride sign code, right? Would you mind doing this one as well?

Comment (IntelOrca) [inline src/interface/paint_path.c]: No that was @duncanspumpkin, I probably just modified it slightly to deal with UTF8 / TTF fonts.

File: src/interface/paint_path.c [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+    } else {
+        regs.ax = height / 16;
+        regs.ah = 0x0A;
+
+    }
+    RCT2_ADDRESS(0x009E3138, uint32)[RCT2_GLOBAL(0x141F56A, uint8) / 2] = regs.eax;
+    RCT2_GLOBAL(0x141F56A, uint8)++;
+}
+
+void viewport_path_paint_setup(uint8 direction, int height, rct_map_element * map_element)
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: Should be 100% done until branching point

File: src/interface/paint_path.c [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+        loc_6A3B57();
+        RCT2_GLOBAL(0x00F3EF6C, uint32) = (uint32) dword_F3EF6C;
+        RCT2_GLOBAL(0x00F3F038, uint16) = word_F3F038;
+        RCT2_GLOBAL(0x00F3EF70, uint16) = dword_F3EF70;
+        RCT2_GLOBAL(0x00F3EF74, uint16) = dword_F3EF74;
+        RCT2_CALLPROC_X(0x6A3B57, 0, 0, get_current_rotation(), height, (int) map_element, 0, 0);
+    }
+}
+
+void loc_6A37C9(rct_map_element * map_element, int height, rct_path_type * dword_F3EF6C, bool word_F3F038, uint32 dword_F3EF70, uint32 dword_F3EF74)
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: Should be 100% done. After checking I'll do the other branch.

File: src/interface/paint_path.c [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+
+    return false;
+}
+
+void loc_6A3FED()
+{
+
+}
+
+bool do_sub_6A2ECC(int supportType, int special, int height, uint32 imageColourFlags, rct_path_type * dword_F3EF6C, bool * underground)
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: Still need to do uneven ground

File: src/interface/paint_path.c [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+/**
+ * rct2: 0x006A3F61
+ * @param map_element (esp[0])
+ * @param bp (bp)
+ * @param height (dx)
+ * @param dword_F3EF6C (0x00F3EF6C)
+ * @param dword_F3EF70 (0x00F3EF70)
+ * @param dword_F3EF74 (0x00F3EF74)
+ */
+void sub_6A3F61(rct_map_element * map_element, uint16 bp, uint16 height, rct_path_type * dword_F3EF6C, uint32 dword_F3EF70, uint32 dword_F3EF74)
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: Need help with this one

File: src/interface/paint_path.c [hunk 6]
```diff
@@ -0,0 +1,775 @@
...
+}
+
+/**
+ * rct2: 0x006A4101
+ * @param map_element (esi)
+ * @param (ecx)
+ * @param ebp (ebp)
+ * @param base_image_id (0x00F3EF78)
+ */
+void sub_6A4101(rct_map_element * map_element, uint16 height, uint32 ebp, bool word_F3F038, rct_path_type * dword_F3EF6C, uint32 base_image_id)
```
Comment (marijnvdwerf) [inline src/interface/paint_path.c]: Should be done, except for sign drawing.

File: src/interface/paint_path.c (hunk 2 above)
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: Add the rct2: address

File: src/interface/paint_path.c [hunk 7]
```diff
@@ -0,0 +1,775 @@
...
+
+void viewport_path_paint_setup(uint8 direction, int height, rct_map_element * map_element)
+{
+    RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8) = VIEWPORT_INTERACTION_ITEM_FOOTPATH;
+
+    uint32 ecx = get_current_rotation();
+    bool word_F3F038 = false;
+
+    uint32 dword_F3EF74 = 0;
+    uint32 dword_F3EF70 = 0;
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: F3EF70 is the path image flags (and colour) and F3EF74 is the path bits (i.e. benches) image flags (and colour)

File: src/interface/paint_path.c [hunk 8]
```diff
@@ -0,0 +1,775 @@
...
+        RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8) = VIEWPORT_INTERACTION_ITEM_NONE;
+        dword_F3EF70 = RCT2_ADDRESS(0x993CC4, uint32_t)[gConfigGeneral.construction_marker_colour];
+    }
+
+    sint16 x = RCT2_GLOBAL(0x009DE56A, sint16), y = RCT2_GLOBAL(0x009DE56E, sint16);
+
+    rct_map_element * surface = map_get_surface_element_at(x / 32, y / 32);
+
+    uint16 bl = height / 8;
+    if (surface == NULL) {
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: There will always be a surface but i guess you should keep this in.

File: src/interface/paint_path.c [hunk 9]
```diff
@@ -0,0 +1,775 @@
...
+        word_F3F038 = true;
+    } else if (surface->base_height != bl) {
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: just do the height / 8 here no need to create the var bl

File: src/interface/paint_path.c [hunk 10]
```diff
@@ -0,0 +1,775 @@
...
+    } else {
+        if (map_element->properties.path.type & 0x04) {
+            // Diagonal path
+
+            if ((surface->properties.surface.slope & 0x1F) != byte_98D800[map_element->properties.path.type & 0x03]) {
+                word_F3F038 = true;
+            }
+        } else {
+            if (surface->properties.surface.slope & 0x1F) {
+                word_F3F038 = true;
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: it seems like `word_F3F038` is the flag stating that you should draw flooring underneath the tile i.e. path in the air/slopped.

File: src/interface/paint_path.c [hunk 11]
```diff
@@ -0,0 +1,775 @@
...
+        uint8 patrolColour = COLOUR_LIGHT_BLUE;
+
+        x = (x & 0x1F80) >> 7;
+        y = (y & 0x1F80) >> 1;
+        int offset = (x | y) >> 5;
+        int bitIndex = (x | y) & 0x1F;
+
+        bool do_it = false;
+
+        if (staffIndex >= 0) {
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: This bit seems very similar to the surface code. I would suggest making a function that does all of this.

File: src/interface/paint_path.c [hunk 12]
```diff
@@ -0,0 +1,775 @@
...
+                height += 16;
+            }
+
+            sub_98196C(imageId | patrolColour << 19 | 0x20000000, 16, 16, 1, 1, 0, height + 2, get_current_rotation());
+        }
+    }
+
+
+    if (gCurrentViewportFlags & VIEWPORT_FLAG_PATH_HEIGHTS) {
+        uint16 height = 3 + map_element->base_height * 8;
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: Missing a check to make sure that this is not zoomed.

File: src/interface/paint_path.c [hunk 13]
```diff
@@ -0,0 +1,775 @@
...
+            height += 8;
+        }
+        uint32 imageId = (SPR_HEIGHT_MARKER_BASE + height / 16) | COLOUR_GREY << 19 | 0x20000000;
+        imageId += RCT2_GLOBAL(RCT2_ADDRESS_CONFIG_HEIGHT_MARKERS, uint16);
+        imageId -= RCT2_GLOBAL(0x01359208, uint16);
+        sub_98196C(imageId, 16, 16, 1, 1, 0, height, get_current_rotation());
+    }
+
+    uint8 pathType = (map_element->properties.path.type & 0xF0) >> 4;
+    rct_path_type * dword_F3EF6C = (rct_path_type *) object_entry_groups[OBJECT_TYPE_PATHS].chunks[pathType];
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: Use `g_pathTypeEntries `

File: src/interface/paint_path.c [hunk 14]
```diff
@@ -0,0 +1,775 @@
...
+
+    if (dword_F3EF6C->var_0A == 0) {
+        loc_6A37C9(map_element, height, dword_F3EF6C, word_F3F038, dword_F3EF70, dword_F3EF74);
+    } else {
+        loc_6A3B57();
+        RCT2_GLOBAL(0x00F3EF6C, uint32) = (uint32) dword_F3EF6C;
+        RCT2_GLOBAL(0x00F3F038, uint16) = word_F3F038;
+        RCT2_GLOBAL(0x00F3EF70, uint16) = dword_F3EF70;
+        RCT2_GLOBAL(0x00F3EF74, uint16) = dword_F3EF74;
+        RCT2_CALLPROC_X(0x6A3B57, 0, 0, get_current_rotation(), height, (int) map_element, 0, 0);
```
Comment (duncanspumpkin) [inline src/interface/paint_path.c]: Can you add a little not stating that this is just for running the old code. I got a little confused wondering why there was a double function call here.

**Comment by janisozaur** [comment]
Please prefix new files with license header.
```
#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
/*****************************************************************************
 * OpenRCT2, an open source clone of Roller Coaster Tycoon 2.
 *
 * OpenRCT2 is the work of many authors, a full list can be found in contributors.md
 * For more information, visit https://github.com/OpenRCT2/OpenRCT2
 *
 * OpenRCT2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * A full copy of the GNU General Public License can be found in licence.txt
 *****************************************************************************/
#pragma endregion
```

**Comment by LinusU** [comment]
@marijnvdwerf I saw that a lot of your other paint work got merged, any update on this one? Keep up the awesome work 🙌

File: src/world/footpath.h [hunk 15]
```diff
@@ -34,8 +34,9 @@ typedef struct rct_footpath_entry {
 	uint32 bridge_image;		// 0x06
 	uint8 var_0A;
 	uint8 flags;				// 0x0B
+	uint8 scrolling_mode;		// 0x0C
 } rct_footpath_entry;
-assert_struct_size(rct_footpath_entry, 12);
+assert_struct_size(rct_footpath_entry, 13);
```
Comment (marijnvdwerf) [inline src/world/footpath.h]: @duncanspumpkin Is this supposed to be an odd number?

Comment (Goddesen) [inline src/world/footpath.h]: 2 + 4 + 4 + 1 + 1 + 1 = 13. Seems fine to me?

Comment (marijnvdwerf) [inline src/world/footpath.h]: The other structs I've come across are usually even numbered with padding at the end. That's why this seems kinda odd.

Comment (Goddesen) [inline src/world/footpath.h]: Ah, I get what you're saying. Should maybe add manual padding for packed arrays then?
As a side note, I thought unaligned memory access wasn't really a problem on x86 anymore, except for certain SIMD instructions?

Comment (duncanspumpkin) [inline src/world/footpath.h]: There is actually another var not used that takes it back to an even number. No where in the code is there an increment of the array so you never actually notice this.

**Pull request merged**